### PR TITLE
s3 backup config check

### PR DESCRIPTION
**Pull request by shaik80**
### :nut_and_bolt: Description: What code changed, and why?

s3/minio backup config check
### :chains: Related Resources
https://chefio.atlassian.net/browse/CHEF-2196
### :athletic_shoe: How to Build and Test the Change
go to `automate/components/automate-cli`
run `make verify-serve  `
go to postman
```
method: post
url: http://localhost:7799/api/v1/checks/s3-config
requestbody:
{
  "endpoint": "s3://example-s3.aws.region.com",
  "bucket_name": "backups",
  "base_path": "automate",
  "access_key": "VALID-ACCESS-KEY",
  "secret_key": "SECRET-KEY"
}
```
### :white_check_mark: Checklist

**All PRs** must tick these:

- [ ] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [ ] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [ ] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [ ] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [ ] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [ ] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### :camera: Screenshots, if applicable
[S3 test](https://progresssoftware-my.sharepoint.com/:v:/g/personal/smudassi_progress_com/ESjy_9Yf5KFNoa91I2fTH_QBgt7HwfNms1Ib6uRQoajAJw?e=uBrd1J)
[Minio test](https://progresssoftware-my.sharepoint.com/:v:/g/personal/smudassi_progress_com/EaEdYteTlSpGu5__H76MXbwBH3OOJEIm_my3g_4tThQD9A?e=stTeYd)
Screenshot 2023-05-22 at 6 20 32 PM
Screenshot 2023-05-22 at 6 20 45 PM
Screenshot 2023-05-22 at 6 21 09 PM

### CODE REVIEW

File: components/automate-cli/pkg/verifyserver/models/s3config.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package models
+
+type ServiceCheck struct {
```
Comment (vivekshankar1) [inline components/automate-cli/pkg/verifyserver/models/s3config.go:3]: Change the name appropriately

File: components/automate-cli/pkg/verifyserver/server/api/v1/s3config_test.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
+package v1_test
```
Comment (vivekshankar1) [inline components/automate-cli/pkg/verifyserver/server/api/v1/s3config_test.go:1]: add more test cases for incorrect request body

**Review by vivekshankar1 — CHANGES_REQUESTED** [review]
Add debug logs to keep track of all steps

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+)
+
+var (
+	s3ConnectionTitle           = "S3 connection test"
+	s3ConnectionErrorMsg        = "Machine is not able to connect with S3 using the provided access key and secret key"
+	s3ConnectionResolutionMsg   = "Provide the correct S3 url or access or secret keys"
+	s3ConnectionSuccessMsg      = "Machine is able to connect with S3 using the provided access key and secret key"
+	s3BucketAccessTitle         = "S3 bucket access test"
+	s3BucketAccessErrorMsg      = "Machine is not able to access the S3 bucket using the provided access key and secret key"
+	s3BucketAccessResolutionMsg = "Provide the necessary acess to the S3 bucket"
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:23]: typo - acess -> access

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	GetS3Connection(models.S3ConfigRequest) models.ServiceCheck
+	GetBucketAccess(models.S3ConfigRequest) models.ServiceCheck
+}
+
+type S3ConfigService struct {
+	logger logger.Logger
+	req    models.S3ConfigRequest
+}
+
+func NewS3ConfigService(logger logger.Logger) S3Config {
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:37]: you should return the struct object not the interface

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	s3ConnectionResolutionMsg   = "Provide the correct S3 url or access or secret keys"
+	s3ConnectionSuccessMsg      = "Machine is able to connect with S3 using the provided access key and secret key"
+	s3BucketAccessTitle         = "S3 bucket access test"
+	s3BucketAccessErrorMsg      = "Machine is not able to access the S3 bucket using the provided access key and secret key"
+	s3BucketAccessResolutionMsg = "Provide the necessary acess to the S3 bucket"
+	s3BucketAccessSuccessMsg    = "Machine is able to access the S3 bucket using the provided access key and secret key"
+)
+
+type S3Config interface {
+	GetS3Connection(models.S3ConfigRequest) models.ServiceCheck
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:28]: methods should take the pointer of S3ConfigRequest not the object

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	GetBucketAccess(models.S3ConfigRequest) models.ServiceCheck
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:29]: methods should take the pointer of S3ConfigRequest, not the object

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+func TestListBuckets(t *testing.T) {
+	log, _ := logger.NewLogger("text", "debug")
+	cs := s3configservice.NewS3ConfigService(log)
+	sess, err := cs.(*s3configservice.S3ConfigService).AwsConnection()
+	assert.NoError(t, err)
+	s3Client := s3.New(sess)
+	err = cs.(*s3configservice.S3ConfigService).ListBuckets(s3Client)
+	assert.Error(t, err)
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:74]: you have tested the errored case, but where is the success case?

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+func TestUploadObject(t *testing.T) {
+	log, _ := logger.NewLogger("text", "debug")
+	cs := s3configservice.NewS3ConfigService(log)
+	sess, err := cs.(*s3configservice.S3ConfigService).AwsConnection()
+	assert.NoError(t, err)
+	// s3Client := s3.New(sess)
+	err = cs.(*s3configservice.S3ConfigService).UploadObject(sess)
+	assert.Error(t, err)
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:84]: you have tested the errored case, but where is the success case?

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+func TestListObjects(t *testing.T) {
+	log, _ := logger.NewLogger("text", "debug")
+	cs := s3configservice.NewS3ConfigService(log)
+	sess, err := cs.(*s3configservice.S3ConfigService).AwsConnection()
+	assert.NoError(t, err)
+	s3Client := s3.New(sess)
+	err = cs.(*s3configservice.S3ConfigService).ListObjects(s3Client)
+	assert.Error(t, err)
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:94]: you have tested the errored case, but where is the success case?

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+func TestDeleteObjects(t *testing.T) {
+	log, _ := logger.NewLogger("text", "debug")
+	cs := s3configservice.NewS3ConfigService(log)
+	sess, err := cs.(*s3configservice.S3ConfigService).AwsConnection()
+	assert.NoError(t, err)
+	s3Client := s3.New(sess)
+	err = cs.(*s3configservice.S3ConfigService).DeleteObjects(s3Client)
+	assert.Error(t, err)
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:104]: you have tested the errored case, but where is the success case?

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
+package s3configservice_test
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:1]: please write the proper unit tests

File: components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+func (h *Handler) GetS3Config(c *fiber.Ctx) {
+	s3ConfigRequest := new(models.S3ConfigRequest)
+	if err := c.BodyParser(&s3ConfigRequest); err != nil {
+		errString := fmt.Sprintf("s3 config request body parsing failed: %v", err.Error())
+		h.Logger.Error(fmt.Errorf(errString))
+		c.Status(fiber.StatusBadRequest).JSON(response.BuildFailedResponse(&fiber.Error{Message: err.Error()}))
+		return
+	}
+	s3Connection := h.S3ConfigService.GetS3Connection(*s3ConfigRequest)
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go:19]: what if the s3 connection fails? when the connection fails, what is the need of checking the bucket access?

**Review by Dmaddu — CHANGES_REQUESTED** [review]

File: components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go (hunk 12 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go:19]: s3 connection just gives session, it doesn’t validate your access key and secret key , to validate secret and access key we need to run list bucket

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go (hunk 11 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:1]: we need to hit actual aws function to get success response, so only failure unit test case is written

File: components/automate-cli/pkg/verifyserver/models/s3config.go (hunk 1 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/models/s3config.go]: Done

File: components/automate-cli/pkg/verifyserver/server/api/v1/s3config_test.go (hunk 2 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/server/api/v1/s3config_test.go:1]: Done

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go (hunk 3 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go]: Done

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go (hunk 7 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:81]: Done

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go (hunk 8 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:91]: Done

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go (hunk 9 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice_test.go:101]: Done, adding success test case

File: components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+	"github.com/chef/automate/components/automate-cli/pkg/verifyserver/response"
+	"github.com/gofiber/fiber"
+)
+
+func (h *Handler) GetS3Config(c *fiber.Ctx) {
+	s3ConfigRequest := new(models.S3ConfigRequest)
+	if err := c.BodyParser(&s3ConfigRequest); err != nil {
+		errString := fmt.Sprintf("s3 config request body parsing failed: %v", err.Error())
+		h.Logger.Error(fmt.Errorf(errString))
+		c.Status(fiber.StatusBadRequest).JSON(response.BuildFailedResponse(&fiber.Error{Message: err.Error()}))
```
Comment (vivekshankar1) [inline components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go:16]:
```suggestion
		c.Next(&fiber.Error{Code: fiber.StatusBadRequest, Message: err.Error()})
```

Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go]: Done

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+		Logger:   logger,
+		AwsUtils: awsUtils,
+	}
+}
+
+func (ss *S3ConfigService) GetS3Connection(req *models.S3ConfigRequest) *models.S3ServiceCheck {
+	ss.Req = req
+	sess, err := ss.AwsConnection()
+	if err != nil {
+		return ss.Response(s3ConnectionTitle, "", errors.Wrap(err, s3ConnectionErrorMsg).Error(), s3ConnectionResolutionMsg, false)
```
Comment (rishabhjhs) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:45]: Add log message for the error

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	s3Client := ss.AwsUtils.New(sess)
+	err = ss.ListBuckets(s3Client)
+	if err != nil {
+		return ss.Response(s3ConnectionTitle, "", errors.Wrap(err, s3ConnectionErrorMsg).Error(), s3ConnectionResolutionMsg, false)
```
Comment (rishabhjhs) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:50]: Add log message

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+func (ss *S3ConfigService) GetBucketAccess(req *models.S3ConfigRequest) *models.S3ServiceCheck {
+	ss.Req = req
+
+	// S3 connection
+	sess, err := ss.AwsConnection()
+	s3Client := ss.AwsUtils.New(sess)
+	if err != nil {
+		return ss.Response(s3BucketAccessTitle, "", errors.Wrap(err, s3BucketAccessErrorMsg).Error(), s3BucketAccessResolutionMsg, false)
```
Comment (rishabhjhs) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:62]: Add logger

**Review by rishabhjhs — APPROVED** [review]

File: components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+func NewAwsUtils() AwsUtils {
+	return &AwsUtilsImpl{}
+}
+
+func (au *AwsUtilsImpl) NewSessionWithOptions(endpoint, accessKey, secretKey string) (*session.Session, error) {
+	config := aws.Config{
+		Endpoint:         &endpoint,
+		Region:           aws.String("us-east-1"),
```
Comment (punitmundra) [inline components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go:32]: why region is hardcoded ?

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 18]
```diff
@@ -0,0 +1,144 @@
+package s3configservice
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:1]: coverage is very less, please add the tests to cover the both positive and negative scenarios

File: components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go (hunk 12 above)
Comment (Dmaddu) [reply components/automate-cli/pkg/verifyserver/server/api/v1/s3config.go]: Here you are returning the check response `S3ServiceCheck`, which has passed field, my question is, in case the passed value is false, do we still need to check the `GetBucketAccess`?

File: components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go [hunk 19]
```diff
@@ -0,0 +1,72 @@
+package awsutils
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go:1]: No tests for this package

**Review by Dmaddu — CHANGES_REQUESTED** [review]

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	// upload data in s3 bucket
+	err = ss.UploadObject(sess)
+	if err != nil {
+		return ss.Response(s3BucketAccessTitle, "", errors.Wrap(err, s3BucketAccessErrorMsg).Error(), s3BucketAccessResolutionMsg, false)
+	}
+
+	// read/list data in s3 bucket
+	err = ss.ListObjects(s3Client)
```
[the remark this answers is not in the document]
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:78]: Done

File: components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go (hunk 17 above)
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go]: now taking region from req body

Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go]: Done

File: components/automate-cli/pkg/verifyserver/server/api/v1/handler.go [hunk 21]
```diff
@@ -2,14 +2,16 @@ package v1
...
 	"github.com/chef/automate/lib/logger"
 )
 
 type Handler struct {
-	Logger        logger.Logger
-	StatusService statusservice.IStatusService
+	Logger          logger.Logger
+	StatusService   statusservice.IStatusService
 	BatchCheckService batchcheckservice.IBatchCheckService
+	S3ConfigService s3configservice.S3Config
```
[the remark this answers is not in the document]
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/server/api/v1/handler.go:14]: Done

**Comment by shaik80** [comment]
working on conflicts

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+	if err != nil {
+		return ss.Response(s3BucketAccessTitle, "", errors.Wrap(err, s3BucketAccessErrorMsg).Error(), s3BucketAccessResolutionMsg, false)
+	}
+	return ss.Response(s3BucketAccessTitle, s3BucketAccessSuccessMsg, "", "", true)
+}
+
+func (ss *S3ConfigService) AwsConnection() (*session.Session, error) {
+	sess, err := ss.AwsUtils.NewSessionWithOptions(ss.Req.Endpoint, ss.Req.AccessKey, ss.Req.SecretKey, ss.Req.Region)
+	if err != nil {
+		ss.Logger.Error("s3 config aws connection failed: ", err.Error())
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:91]: No unit test for this. If you could have written the tests for above mentioned scenarios, this will cover on that.

**Review by Dmaddu — CHANGES_REQUESTED** [review]

File: components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+var (
+	s3ConnectionTitle           = "S3 connection test"
+	s3ConnectionErrorMsg        = "Machine is not able to connect with S3 using the provided access key and secret key"
+	s3ConnectionResolutionMsg   = "Provide the correct S3 url or access or secret keys"
+	s3ConnectionSuccessMsg      = "Machine is able to connect with S3 using the provided access key and secret key"
+	s3BucketAccessTitle         = "S3 bucket access test"
+	s3BucketAccessErrorMsg      = "Machine is not able to access the S3 bucket using the provided access key and secret key"
+	s3BucketAccessResolutionMsg = "Provide the necessary access to the S3 bucket"
+	s3BucketAccessSuccessMsg    = "Machine is able to access the S3 bucket using the provided access key and secret key"
+	region                      = ""
```
[the remark this answers is not in the document]
Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/services/s3configservice/s3configservice.go:21]: yes this is needed, i need to check list buck to check S3 connection, that can be only done by using default region, i am storing region in  global variable and then making region as empty to list buck ( test connection ), then using global variable to access bucket to upload, delete and list object (S3 bucket can be region specific)

**Review by atultherajput — COMMENTED** [review]
Can we improve code coverage?

**Review by vivekshankar1 — APPROVED** [review]

File: components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func NewAwsUtils() AwsUtils {
+	return &AwsUtilsImpl{}
+}
+
+func (au *AwsUtilsImpl) NewSessionWithOptions(endpoint, accessKey, secretKey, region string) (*session.Session, error) {
+	var config aws.Config
+	if region == "" {
+		region = "us-east-1"
```
Comment (punitmundra) [inline components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go:32]: if region id empty (config.toml), and bucket is in some other region other than us-east-1, with it work or throws error ?

Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go:32]: if you give region as empty, i am assigning Default region **us-east-1**, if bucket is in another region its mandatory to give

Comment (shaik80) [reply components/automate-cli/pkg/verifyserver/utils/awsutils/aws.go:32]: we don’t need region in minio, so in this case we need to assign default region to get aws session

**Review by Dmaddu — APPROVED** [review]

**Review by punitmundra — APPROVED** [review]
